### ISSUE TITLE
[Feature request]: Special Action workflow for PRs

### ISSUE THREAD

**Issue by DonKoko** [Shelf-nu/shelf.nu#319]
### Is your feature request related to a problem? Please describe?

Currently when someone opens a PR to main from a fork, the e2e tests fail because of the required env variables. Here is the only context I could find: https://stackoverflow.com/questions/75873833/how-to-protect-github-secrets-in-pull-request-actions-from-malicious-pull-reques

### Describe the solution you'd like

So ideally, the e2e tests can run, however I am not sure if that's possible. We have the required secrets in the shelf repository but from what I read those are not passed onto PRs running the tests.

Someone mentioned to me that we can do a workflow that runs instead of the default one when a PR is opened from a fork. Then we can run all the steps from our deploy workflow except the e2e tests. 

In any case this needs to be looked into and any suggestions how to solve this are welcome.

**Comment by kunal00000** [comment]
@DonKoko, I read on internet that we can use `on: pull_request_target` but this comes with great dangers as this gives the PR author the ability to access secrets and change the repo contents.

**Comment by DonKoko** [comment]
hey @kunal00000
Honestly at first when I faced this issue, I was thinking that the best would be if they can place the secrets on their fork and then when the PR happens the tests would run using those secrets. This was of course just my imagination so I am not sure if it works.

As for the privacy issues, you are correct. I heard something like this from someone else and this is indeed a bit dangerous. That's why I was wondering if we can make a workflow that runs instead of the default one when a PR is merged is opened to main but skip e2e tests. They will anyways run on merge so even if something is broken we will catch it. 
I also got the idea from someone else:
> The best thing to do, i guess, is to run a docker with supabase in the workflow 😅

I am not sure how we can do this but I am open to any suggestions.

**Comment by kunal00000** [comment]
What if the workflow needs permission of some authentic users for it to run?
Then I think it's secure.

reference: https://iterative.ai/blog/testing-external-contributions-using-github-actions-secrets

**Comment by DonKoko** [comment]
> What if the workflow needs permission of some authentic users for it to run? Then I think it's secure as this way reviewers can confirm that PR is not trying to access secrets.
> 
> reference: https://iterative.ai/blog/testing-external-contributions-using-github-actions-secrets

Article looks very promising. I will read it as soon as I can and let you know what I think.

**Comment by SuperSchek** [comment]
What's up broski's!

You could solve this by requiring approval for workflow runs from forks. That way people can fork to their hearts content and you benevolent dictators can trigger a run when you review the PR. Many public repositories handle it this way.

More info on this in the GH docs: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

**Comment by carlosvirreira** [comment]
@SuperSchek thanks for your comment man! Glad to see your beautiful face here <3

**Comment by DonKoko** [comment]
Hey @SuperSchek . Thanks for the input. Great to have you here. 
So indeed its the same thing that the article shared by @kunal00000 suggests. I think it just needs to be configured because even now when there is a PR from a fork, i need to approve for the tests to run but they fail because of the issue explained above so I guess if we do it as in the example, the secrets will be passed on so it will work.

@kunal00000 would you be interested in giving the bounty a go?

**Comment by kunal00000** [comment]
Yes, I will be happy to take this up.

So my plan is to create a new .yml file for playwright workflow and remove it from deploy.yml

/attempt #319

**Comment by kunal00000** [comment]
Update: `Test PR from Fork` workflow is working as expected (passing secrets to workflows).
One thing I am not sure about is seeking of permission from external environment users.

**Comment by DonKoko** [comment]
Thanks @kunal00000. I will look at it to try to resolve it but I am pretty certain thats not the exact issue. This ran normally before changing the tests and how the variables are setup.
I will look into it and let you know.

**Comment by kunal00000** [comment]
This is the problem many people encountered after setting up GitHub CI.

test.slow() worked for some. Maybe that works.

Or It may be something else that's causing this.

**Comment by DonKoko** [comment]
@kunal00000 Yes, its something else. I am solving it now. Okey so to close this I just need to understand one last part, how does the `authorize` step work. Like how do I autorize a user so their PR gets the tests running?

**Comment by kunal00000** [comment]
@DonKoko . Yes, to approve PR for running tests you will have to add yourself as required reviewers for `external` environment. 

So all the forks making PR are basically deploying on an environment and if there are required reviewers then without their approval PR cannot be deployed.

**Comment by DonKoko** [comment]
Okey and once I create an environment, do we need to change the configuration of the workflows? Do I have to add the variables to that environment in the settings of the repo or those have to stay as they are now?

**Comment by DonKoko** [comment]
<img width="806" alt="Screenshot 2023-09-08 at 17 33 51" src="https://github.com/Shelf-nu/shelf.nu/assets/7840007/34a52022-8b45-470c-a661-52e038f390c3">
@kunal00000 we have an environment called external already present there. I added myself to required reviewers. How can we now test this if the vars are being passed, because in the screenshot you can see there is a place to set secrets per environment.

**Comment by kunal00000** [comment]
Yes making a test PR

**Comment by DonKoko** [comment]
@kunal00000 its done. Unfortunately I have to leave in a few minutes as I have some personal arrangements and I wont be able to look at it until monday agian.

**Comment by kunal00000** [comment]
Okay no worries 👍🏻.

**Comment by kunal00000** [comment]
>  I hope if I have a question or 2 next week you can jump in quickly to help.

Yes sure happy to help. thanks for the tip.

**Comment by carlosvirreira** [comment]
@kunal00000 l, would love to have you on our discord!

https://discord.gg/2fNUw7DT

**Comment by DonKoko** [comment]
@kunal00000 we have the issues with the tests still but for some reason for 1 day the test worked. Now that I know it works I will close the issue. Thanks again.
